### PR TITLE
Feature/hub 374 refer to other articles from article

### PR DESCRIPTION
**Pull request by janipalsamaki**
# See also Instructions for students

- Adds support for referring to articles from an article.
- This functionality is restricted by domain.
- The functionality is available only when the active domain is the teacher domain.
- The reference widget displays articles from student domain only.
- Introduces a new module that will provide domain-related services to other modules (`uhsg_domain`).
- Sets the domain for an article on presave based on user group.
- The domain field is hidden and should not be available for the end user to manipulate.
- Fixes some minor visual regressions that are also present in `master`.

## Testing tips

- After build, run this four times (wait a bit in between runs to allow for Solr integration to complete indexing): `drush uhsg-assign-article-domains 100`

- One needs to either have both the student domain and the teacher domain configured or temporarily modify `uhsg_domain_entity_field_access` to allow access to the reference field to see the field in edit and view modes.

- Save at least a couple of articles using one of the student user groups. These will then appear in the reference widget (that has domain filtering enabled). Note that the field is translation-specific.

### CODE REVIEW

File: modules/uhsg_domain/uhsg_domain.info.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+name: UHSG Domain
+type: module
+description: Provides domain functionality for other modules.
```
Comment (mikaelkundert) [inline modules/uhsg_domain/uhsg_domain.info.yml]: "Domain functionality" doesn't explain that much, could we try something like:
> Provides support for creating guide-content for student and teacher domains.

Same thing would go for `hook_help()`.

File: modules/uhsg_domain/uhsg_domain.module [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    //$entity->field_domain_source = $domain_id;
+  }
+}
+
+/**
+ * @return DomainService
+ */
+function uhsg_domain_get_service() {
+  return Drupal::service('uhsg_domain.domain');
+}
```
Comment (mikaelkundert) [inline modules/uhsg_domain/uhsg_domain.module]: I see only one use case for this (`uhsg_domain_entity_field_access()`)?

Can we skip using this abstraction and just use directly `\Drupal::service('uhsg_domain.domain')`?

I personally would even use that even I would have more than one use case.

File: modules/uhsg_domain/uhsg_domain.module [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * Sets article domain based on the selected user group. Teacher domain when
+ * the user group is teachers, student domain for other user groups.
+ */
+function uhsg_domain_entity_presave(Drupal\Core\Entity\EntityInterface $entity) {
+  if ($entity instanceof \Drupal\node\Entity\Node && $entity->getType() == 'article') {
+    $domain_id = $entity->field_user_group->value == 'teachers' ? DomainService::TEACHER_DOMAIN_ID : DomainService::STUDENT_DOMAIN_ID;
+    $entity->field_article_domain = $domain_id;
+    // TODO: Set domain source. Both domains need to be accessible for this to
+    // work, since the links will be rewritten to point to one of the domains.
+    //$entity->field_domain_source = $domain_id;
```
[the remark this answers is not in the document]
Comment (janipalsamaki) [reply modules/uhsg_domain/uhsg_domain.module]: We can not set the domain source until the teacher domain can be accessed in production (see the TODO comment).

File: modules/uhsg_domain/uhsg_domain.module [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\uhsg_domain\DomainService;
+
+/**
+ * Implements hook_help().
+ */
+function uhsg_domain_help($route_name, RouteMatchInterface $route_match) {
+  switch ($route_name) {
+    // Main module help for the uhsg_domain module.
+    case 'help.page.uhsg_article':
```
[the remark this answers is not in the document]
Comment (janipalsamaki) [reply modules/uhsg_domain/uhsg_domain.module]: Fixed.

File: modules/uhsg_domain/uhsg_domain.info.yml (hunk 1 above)
Comment (janipalsamaki) [reply modules/uhsg_domain/uhsg_domain.info.yml]: Tuned.

File: modules/uhsg_domain/uhsg_domain.module (hunk 2 above)
Comment (janipalsamaki) [reply modules/uhsg_domain/uhsg_domain.module]: One benefit of the function is to provide type-hinting to IDEs. We can probably manage without, so removed the abstraction.

**Pull request merged**